extensions/data-plane/data-plane-google-bigquery/src/main/java/org/eclipse/edc/connector/dataplane/gcp/bigquery/pipeline/BigQueryDataSink.java

```diff
@@ -0,0 +1,337 @@
+        }
+
+        return success();
+    }
+
+    private void checkStreamWriter()
+        throws DescriptorValidationException, IOException, InterruptedException {
+        if (streamWriter == null) {
+            monitor.debug("BigQuery Sink creating stream writer in pending mode for table " + target.getTableName().toString());
+            var stream = WriteStream.newBuilder()
+                    .setType(WriteStream.Type.PENDING)
+                    .build();
+
+            var createWriteStreamRequest = CreateWriteStreamRequest.newBuilder()
+                    .setParent(target.getTableName().toString())
+                    .setWriteStream(stream)
+                    .build();
+
+            var writeStream = writeClient.createWriteStream(createWriteStreamRequest);
+
+            var builder = JsonStreamWriter.newBuilder(writeStream.getName(), writeClient);
+
+            if (configuration.rpcEndpoint() != null) {
+                builder.setCredentialsProvider(NoCredentialsProvider.create());
+            } else {
+                builder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+            }
+
+            streamWriter = builder.build();
+        }
+
+        if (streamWriter.isClosed()) {
+            // TODO re-create the writer for a maximum, specified number of times.
+            monitor.info("BigQuery Sink stream writer closed, recreating it for stream " + streamWriter.getStreamName());
+            var builder = JsonStreamWriter.newBuilder(streamWriter.getStreamName(), writeClient);
+            if (configuration.rpcEndpoint() != null) {
+                builder.setCredentialsProvider(NoCredentialsProvider.create());
+            } else {
+                builder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+            }
+            streamWriter = builder.build();
+        }
+    }
+
+    private void initService() throws IOException {
+        if (writeClient != null) {
+            return;
+        }
+
+        var settingsBuilder = BigQueryWriteSettings.newBuilder();
+        var host = configuration.rpcEndpoint();
+        if (host != null) {
+            var index = host.indexOf("//");
+            if (index != -1) {
+                host = host.substring(index + 2);
+            }
+            settingsBuilder.setEndpoint(host);
+            settingsBuilder.setTransportChannelProvider(
+                    FixedTransportChannelProvider.create(
+                        GrpcTransportChannel.create(
+                            NettyChannelBuilder.forTarget(host).usePlaintext().build()
+                        )
+                    )
+            );
+
+            settingsBuilder.setCredentialsProvider(NoCredentialsProvider.create());
+        } else {
+            settingsBuilder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+        }
+        writeClient = BigQueryWriteClient.create(settingsBuilder.build());
```

Comment:
I think the SDK class `writeClient` should be externally created and passed into the BQ sink rather than be instantiated directly here, to decouple functionalities and improve testability.

Reply:
Done